client/app/bundles/course/reference-timelines/contexts/LastSavedContext.tsx

```diff
@@ -0,0 +1,59 @@
+import {
+  createContext,
+  Dispatch,
+  ReactNode,
+  SetStateAction,
+  useContext,
+  useState,
+} from 'react';
+import moment from 'moment';
+
+type FetchStatus = 'loading' | 'success' | 'failure';
+
+interface LastSavedState {
+  status?: FetchStatus;
+  lastSaved?: moment.Moment;
+}
+
+interface LastSavedUpdater {
+  abortLoading: () => void;
+  startLoading: () => void;
+  setLastSavedToNow: () => void;
+}
+
+type LastSavedSetter = Dispatch<SetStateAction<LastSavedState>>;
+
+const LastSavedStateContext = createContext<LastSavedState>({});
+const LastSavedSetterContext = createContext<LastSavedSetter>(() => {});
+
+interface LoadingProviderProps {
+  children: ReactNode;
+}
+
+export const LastSavedProvider = (props: LoadingProviderProps): JSX.Element => {
+  const [lastSavedState, setLastSavedState] = useState({});
+
+  return (
+    <LastSavedStateContext.Provider value={lastSavedState}>
+      <LastSavedSetterContext.Provider value={setLastSavedState}>
+        {props.children}
+      </LastSavedSetterContext.Provider>
+    </LastSavedStateContext.Provider>
+  );
+};
+
+export const useLastSaved = (): LastSavedState =>
+  useContext(LastSavedStateContext);
+
+export const useSetLastSaved = (): LastSavedUpdater => {
+  const setLastSaved = useContext(LastSavedSetterContext);
+
+  return {
+    abortLoading: () =>
+      setLastSaved((state) => ({ ...state, status: 'failure' })),
```

Comment:
They are already constants, though(?)

With the way `FetchStatus` is defined, the `status` attribute will always conform to either one of these three values, otherwise TypeScript will raise an error. It is essentially an enum.

Reply:
Noted!